ostd/src/arch/x86/cpu/local.rs

```diff
@@ -0,0 +1,98 @@
+///    and is not concurrently accessed for other purposes.
+///  - The caller must ensure that `addr` is a valid address and properly aligned, as required by the CPU.
+///  - This function should only be called in contexts where the CPU is in a state to accept such changes,
+///    such as during processor initialization.
+pub(crate) unsafe fn set_base(addr: u64) {
+    FS::write_base(x86_64::addr::VirtAddr::new(addr));
+}
+
+/// Gets the base address for the CPU local storage by reading the FS base model-specific register.
+pub(crate) fn get_base() -> u64 {
+    FS::read_base().as_u64()
+}
+
+pub mod preempt_lock_count {
+    //! We need to atomically increment and decrement the per-CPU preemption
+    //! lock count, requiring the implementation using a single instruction.
+    //! This requirement is stated by [`crate::task::processor::PreemptInfo`].
+
+    /// Atomically increments the per-CPU preemption lock count.
+    pub(crate) fn inc() {
+        debug_assert_initialized();
+
+        // SAFETY: The inline assembly increments the lock count atomically
+        // without side effects.
+        unsafe {
+            core::arch::asm!(
+                "inc dword ptr fs:[{0}]",
+                in(reg) offset_of_count(),
+                options(nostack),
+            );
+        }
+    }
+
+    /// Atomically decrements the per-CPU preemption lock count.
+    pub(crate) fn dec() {
+        debug_assert_initialized();
+
+        // SAFETY: The inline assembly decrements the lock count atomically
+        // without side effects.
+        unsafe {
+            core::arch::asm!(
+                "dec dword ptr fs:[{0}]",
+                in(reg) offset_of_count(),
+                options(nostack),
+            );
+        }
+    }
+
+    /// Atomically gets the per-CPU preemption lock count.
+    pub(crate) fn get() -> u32 {
+        debug_assert_initialized();
+
+        let count: u32;
+        // SAFETY: The inline assembly reads the lock count atomically
+        // without side effects.
+        unsafe {
+            core::arch::asm!(
+                "mov {0:e}, fs:[{1}]",
+                out(reg) count,
+                in(reg) offset_of_count(),
+                options(nostack),
+            );
+        }
+        count
+    }
+
+    extern "C" {
+        // The symbol is defined in the linker script.
+        //
+        // It points to 32 bits of memory available for preemption count.
+        fn __cpu_local_preempt_lock_count();
+
+        fn __cpu_local_start();
+    }
+
+    fn offset_of_count() -> usize {
+        __cpu_local_preempt_lock_count as *const u32 as usize
+            - __cpu_local_start as *const u32 as usize
+    }
+
+    /// The GDT ensures that the FS segment is initialized to zero on boot.
+    /// This assertion checks that the base address has been set.
+    fn debug_assert_initialized() {
+        debug_assert_ne!(super::get_base(), 0);
```

Comment:
Perhaps it's better to manually add another `#[cfg(debug_assertions)]`? I'm not sure whether this `super::get_base` can be optimized away when the debug assertions are disabled.